### PR TITLE
[ASTextNode2] Add initial implementation for link handling.

### PR DESCRIPTION
**Pull request by appleguy**
This is a fairly basic first step to achieving feature parity between ASTextNode2
and ASTextNode. It does not yet do the 9-box detection of links, and there is
other code from ASTextNode that could be shared to improve this.

However, in the interest of getting a shippable implementation running as soon
as possible, I'm hoping to quickly refine this. Then we can continue improving
it based on the original ASTextNode with the benefit of testing UI against it.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Private/ASTextNode2.mm [hunk 1]
```diff
@@ -488,7 +488,54 @@ - (id)_linkAttributeValueAtPoint:(CGPoint)point
...
+  // truncation token.
+  if (![layout textRangeAtPoint:point]) {
+    *inAdditionalTruncationMessageOut = YES;
+    return nil;
+  }
+
+  NSRange effectiveRange = NSMakeRange(0, 0);
+  for (NSString *attributeName in self.linkAttributeNames) {
+    id value = [self.attributedText attribute:attributeName atIndex:range.start.offset longestEffectiveRange:&effectiveRange inRange:clampedRange];
+    NSString *name = attributeName;
```
Comment (Adlai-Holler) [inline Source/Private/ASTextNode2.mm]: Nit: remove extra local variable – a little confusing.

Comment (Adlai-Holler) [inline Source/Private/ASTextNode2.mm]: Nit: remove extra local variable or make the difference between `name` and `attributeName` more clear.

File: Source/Private/ASTextNode2.mm [hunk 2]
```diff
@@ -488,7 +488,54 @@ - (id)_linkAttributeValueAtPoint:(CGPoint)point
    inAdditionalTruncationMessage:(out BOOL *)inAdditionalTruncationMessageOut
                  forHighlighting:(BOOL)highlighting
 {
-  AS_TEXT_ALERT_UNIMPLEMENTED_FEATURE();
+  ASDN::MutexLocker l(__instanceLock__);
+
+  ASTextContainer *containerCopy = [_textContainer copy];
+  containerCopy.size = self.calculatedSize;
+  ASTextLayout *layout = [ASTextNode2 compatibleLayoutWithContainer:containerCopy text:_attributedText];
```
Comment (Adlai-Holler) [inline Source/Private/ASTextNode2.mm]: I don't think it's required for us to copy the container here (or below at line 740). Except in extreme edge cases, `_textContainer` will be at the displayed size. Is this just extra safety, or was this necessary to make it work?

File: Source/Private/ASTextNode2.mm [hunk 3]
```diff
@@ -488,7 +488,54 @@ - (id)_linkAttributeValueAtPoint:(CGPoint)point
...
+        && [_delegate respondsToSelector:@selector(textNode:shouldHighlightLinkAttribute:value:atPoint:)]
+        && ![_delegate textNode:(ASTextNode *)self shouldHighlightLinkAttribute:name value:value atPoint:point]) {
+      value = nil;
+      name = nil;
+    }
+
+    if (value != nil || name != nil) {
+      *rangeOut = effectiveRange;
+      if (NSMaxRange(*rangeOut) > NSMaxRange(visibleRange)) {
+        (*rangeOut).length = MAX(NSMaxRange(visibleRange) - (*rangeOut).location, 0);
```
Comment (Adlai-Holler) [inline Source/Private/ASTextNode2.mm]: Is it reasonable for this just to be `*rangeOut = NSIntersectionRange(effectiveRange, visibleRange)` or am I missing something?

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Private/ASTextNode2.mm (hunk 2 above)
Comment (appleguy) [inline Source/Private/ASTextNode2.mm]: @Adlai-Holler unfortunately, without these, link handling does not work. I could try removing just one or the other, but instead I think it is going to require a more detailed look at what is going on.

File: Source/Private/ASTextNode2.mm (hunk 1 above)
Comment (appleguy) [inline Source/Private/ASTextNode2.mm]: Done. Doing this required adding a semi-ugly __strong to the iteration variable declaration, as the compiler does generate an error for writing to that variable without it.

File: Source/Private/ASTextNode2.mm (hunk 3 above)
Comment (appleguy) [inline Source/Private/ASTextNode2.mm]: Good suggestion, done.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Private/ASTextNode2.mm (hunk 2 above)
Comment (Adlai-Holler) [inline Source/Private/ASTextNode2.mm]: OK would you be willing to add a comment linking to this discussion?

**Pull request merged**